### DISCUSSION TITLE
Virtualized vs non-virtualized CI macOS environments

### DISCUSSION THREAD

**Discussion by pepicrft** [MobileNativeFoundation/discussions#68] [General]
Hey there!
I'm curious about what's others' stance on virtualizing macOS environments for running CI builds. 

At @shopify, we've been early adopters of Anka and we've been very happy with the technology. The performance is great, and the time to start images from a suspended state is very fast, around a few seconds. Apart from being able to run builds in clean environments, and thus have better determinism, we can use the same abstraction on top of Docker and Anka. Our hosts have a software, Shopify Build, that takes care of parsing and executing a pipeline. Thanks to Anka we don't have to do any pre or post-build steps to clean the environment, neither re-provisioning the hosts periodically. Our hosts are 2 major macOS versions behind and they are able to run Anka images with Big Sur. Because the maintenance cost is very low, we don't need any CI infra team dedicated to this. The same team responsible for mobile tooling takes care of the infrastructure.

What's your preference and why? How is your project's CI setup?

**Comment by JoeSSS** [comment]
We use the same setup at XING. Actually, it was inspired by the Shopify presentation back in the days :)

**Comment by keith** [comment]
There's some talk about VMs or not on https://github.com/MobileNativeFoundation/discussions/discussions/23, but I think this more targeted discussion is good too!

At Lyft we've been using bare metal for the last few years entirely based on all the horror stories we hear about VM performance, and the overall maintenance cost with them in general. We've generally been able to keep the environment "clean enough" that pollution hasn't been a problem (more on that here https://github.com/MobileNativeFoundation/discussions/discussions/23#discussioncomment-456242)

Do you have any numbers you can share on the difference in performance for your case with Anka?

**Reply by raven** [reply]
At realestate.com.au we use Anka and so far have been very happy with [the setup](https://github.com/MobileNativeFoundation/discussions/discussions/23#discussioncomment-500126). We have definitely come across things that significantly improved performance for us along the way though. 
Things such as:
 - When completing the different steps of the vm image building phase, giving the image 'rest' time. Even with disabling as many system services (spotlight etc) we noticed it took time for the system as analysed through the console to return to an optimal idle'ish state. We have these idle times after macOS installation, but also after installing Xcode and it's initial launch. 
 - The iOS simulators we pre-install and pre-launch and then quit + idle time. If a simulator device has never been launched before we have noticed its initial launch and disk allocation significantly hampered system performance even after the simulator was 'ready' and accepting jobs. 
 - Minimal read/writes to the host from disk mounts. We mount a pre-flight directory that we copy from host to VM a set of secrets/env files, but after that there are no reads or writes back to the host. This may be Anka specific, but we found mounted I/O to be very poor performing.

**Comment by raven** [comment]
I'm really keen to hear if anyone using just metal and has experimented with APFS snapshots as a method to ensure consistent environmental state. I think I saw this mentioned by Peter Steinberger somewhere as an idea and it sounded fascinating.

**Comment by ob** [comment]
We run on bare metal and went "single Xcode" meaning only one version of Xcode is installed on each machine. The jobs get routed to different machines if you need different Xcode versions and we only keep a very small set of versions. In our experience, building is not terribly slower in a VM, however testing with parallels simulators takes a big performance hit.

**Reply by shepting** [reply]
We are currently doing bare metal on 250 machine just like this as well (single Xcode per machine). We may be moving this quarter to multiple Xcode but are currently pretty happy with the performance.

**Reply by pepicrft** [reply]
Out of curiosity, how do you route your builds based on the Xcode version specified in the pipeline? Is that something that your CI provider/tool supports?

**Reply by keith** [reply]
We've looked into this in general with buildkite and for our use case we can handle this by swapping Xcode versions based on the current checkout, some example of that lives here too https://github.com/bazelbuild/continuous-integration/blob/2d4945210d7662f9cc287a1acdd510fb99426a07/buildkite/bazelci.py#L1249-L1253

**Reply by raven** [reply]
We have the toolchain declared as part of the Buildkite pipeline. For us, that is captured by our proxy Buildkite agent that sits on metal, the bootstrap hook inspects the pipeline inputs and then grabs the associated VM image, and injects the build step into the internal to VM Buildkite agent. If we needed to farm to different metal, it would be a similar approach of using a proxy Buildkite agent (wouldn't need to be a macOS agent), that would intercept the build and then pass into the appropriate metal agent. 

https://buildkite.com/docs/agent/v3/configuration `bootstrap-script`

![image](https://user-images.githubusercontent.com/316931/113066139-bd19ab80-9205-11eb-9dd9-ed2080cff19a.png)

**Reply by ob** [reply]
> Out of curiosity, how do you route your builds based on the Xcode version specified in the pipeline? Is that something that your CI provider/tool supports?

We have a custom CI system that allows us to route to different pools of machines based on attributes like Xcode version.

**Comment by fkorotkov** [comment]
We also were a pretty early adopters of Anka on Intel but eventually we faced a problem of distributing new versions of templates without service interruption. Anka registry was just not able to single handedly distribute templates to dozens on hosts. So we wrote some automations to manually and in parallel pre-fetch new templates to hosts before switching versions.

Then at in the beginning of 2020 we switched to Parallels because Anka's network was not reliable. Every couple of weeks one of our hosts was getting in a weird state when VMs had flaky network and we need to reset the host. We tried to debug the issue with Veetru team but with no luck. The switch was smooth: we had our Packer templates, slightly modified automations that pre-fetched VMs to work with Parallels, etc. In the end we had a better performance, stability and Parallels licensing cost was a fraction comparing to Anka.

Fast forward to 2022 when we needed to support to Apple Silicon. We ended up building our own free and open sources virtualization solution called [Tart](https://github.com/cirruslabs/tart) based on Apple's `Virtualizatoin.Framework`. One of the key features we did is a support for OCI registries for pushing/pulling VMs from. It allowed to get rig of our hacky scripts for prefetching VMs because nowadays Docker registries are super scalable.

**Reply by ob** [reply]
`tart` looks very interesting, thank you for posting... do you have any performance numbers w.r.t. running iOS tests in parallel? We benchmarked Anka vs bare metal a while ago and while the overhead of VMs during build time was acceptable, the performance of running multiple iOS simulators in parallel for testing was not.

**Reply by fkorotkov** [reply]
Yeah, Anka for Intel had to implement a lot of things including the GPU subsystem which is important for running simulators. In case of Tart and Anka 3.0 for M1 they both built on top of Apple's own `Virtualization.Framework` and Apple did some incredible job stitching hardware and software parts. 

We only ran a few syntactic benchmarks like [Geekbench which showed there is almost no overhead of virtualization layer](https://browser.geekbench.com/v5/cpu/compare/14966395?baseline=14966339). In real workloads VMs performed extremely good as well. There was only one issue with Monterey VMs where the paravirtualized GPU was crashing but I talked to Apple folks during WWDC and they suggested to use Ventura. And indeed Ventura VMs even on Monterey hosts didn't have the crashes anymore.

So I'd suggest you to give another try to either Tart or Anka 3.0 and check out performance for your use case 😉
